xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc

```diff
@@ -526,6 +526,338 @@ class CholeskyOpLowering : public OpRewritePattern<CholeskyOp> {
+              "unexpected number of operands for fused dot attention - "
+              "BMM_Bias_Mask_Softmax_Dropout_BMM");
+        break;
+
+      case mlir::lmhlo_gpu::FusedMhaDagSignature::ScaleMaskSoftmax:
+        if (num_operands == 6) {
+          fused_attention += "scale.mask.softmax.inference";
+        } else if (num_operands == 7) {
+          fused_attention += "scale.mask.softmax.forward";
+        } else
+          return op.emitOpError(
+              "unexpected number of operands for fused dot attention - "
+              "BMM_mask_Softmax_BMM");
+        break;
+
+      case mlir::lmhlo_gpu::FusedMhaDagSignature::ScaleMaskSoftmaxDropout:
+        if (num_operands == 6) {
+          fused_attention += "scale.mask.softmax.dropout.inference";
+        } else if (num_operands == 7) {
+          fused_attention += "scale.mask.softmax.dropout.forward";
+        } else
+          return op.emitOpError(
+              "unexpected number of operands for fused dot attention - "
+              "BMM_mask_Softmax_Dropout_BMM");
+        break;
+
+      case mlir::lmhlo_gpu::FusedMhaDagSignature::ScaleBiasSoftmax:
+        if (num_operands == 6) {
+          fused_attention += "scale.bias.softmax.inference";
+        } else if (num_operands == 7) {
+          fused_attention += "scale.bias.softmax.forward";
+        } else
+          return op.emitOpError(
+              "unexpected number of operands for fused dot attention - "
+              "BMM_bias_Softmax_BMM");
+        break;
+
+      case mlir::lmhlo_gpu::FusedMhaDagSignature::ScaleBiasSoftmaxDropout:
+        if (num_operands == 6) {
+          fused_attention += "scale.bias.softmax.dropout.inference";
+        } else if (num_operands == 7) {
+          fused_attention += "scale.bias.softmax.dropout.forward";
+        } else
+          return op.emitOpError(
+              "unexpected number of operands for fused dot attention - "
+              "BMM_bias_Softmax_Dropout_BMM");
+        break;
+
+      default:
+        return op.emitOpError("Undefined fused dot attention DAG signature");
+    }
+
+    // Get or create a custom call function declaration.
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+    func::FuncOp callee = custom_calls_.GetOrCreate(b, fused_attention, op);
+
+    // Convert fused_attention to a function call.
+    auto call = rewriter.create<func::CallOp>(op.getLoc(), callee.getName(),
+                                              TypeRange(), op.getOperands());
+
+    // Assign a unique id to this instance of a fused_attention operation.
+    call->setAttr(b.getStringAttr("uid"), b.getI64IntegerAttr(uid_.uid()));
+
+    // Helper functins to copy attributes from the conv op to the custom call.
+    auto set_attr = [&](StringRef name, Attribute attr) {
+      call->setAttr(b.getStringAttr(name), attr);
+    };
+
+    set_attr("fmha_scale", op.getFmhaScaleAttr());
+    if (op.getDropoutRate()) {
+      set_attr("dropout_rate", op.getDropoutRateAttr());
```

Comment:
You can add `if (attr) ...` check to `set_attr` and remove some of the branches here, should be more dense code and easier to read